src/Development/IDE/Core/FileExists.hs

```diff
@@ -0,0 +1,161 @@
+modifyFileExistsAction :: (FileExistsMap -> IO FileExistsMap) -> Action ()
+modifyFileExistsAction f = do
+  FileExistsMapVar var <- getIdeGlobalAction
+  liftIO $ modifyVar_ var f
+
+-- | Modify the global store of file exists
+modifyFileExists :: IdeState -> [(NormalizedFilePath, Bool)] -> IO ()
+modifyFileExists state changes = do
+  FileExistsMapVar var <- getIdeGlobalState state
+  changesMap           <- evaluate $ Map.fromList changes
+  modifyVar_ var $ evaluate . Map.union changesMap
+  let flushPreviousValues = do
+        ShakeExtras { state } <- getShakeExtras
+        liftIO $ mapM_ (resetValue state GetFileExists . fst) changes
+  void $ shakeRun state [flushPreviousValues]
+
+-------------------------------------------------------------------------------------
+
+type instance RuleResult GetFileExists = Bool
+
+data GetFileExists = GetFileExists
+    deriving (Eq, Show, Typeable, Generic)
+
+instance NFData   GetFileExists
+instance Hashable GetFileExists
+instance Binary   GetFileExists
+
+-- | Returns True if the file exists
+getFileExists :: NormalizedFilePath -> Action Bool
+getFileExists fp = use_ GetFileExists fp
+
+-- | Installs the 'getFileExists' rules.
+--   Provides a fast implementation if client supports dynamic watched files.
+--   Creates a global state as a side effect in that case.
+fileExistsRules :: ClientCapabilities -> (NormalizedFilePath -> Action Bool) -> Rules ()
+fileExistsRules ClientCapabilities{_workspace}
+  | Just WorkspaceClientCapabilities{_didChangeWatchedFiles} <- _workspace
+  , Just DidChangeWatchedFilesClientCapabilities{_dynamicRegistration} <- _didChangeWatchedFiles
+  , Just True <- _dynamicRegistration
+  = fileExistsRulesFast
+  | otherwise = fileExistsRulesSlow
+
+--   Requires an lsp client that provides WatchedFiles notifications.
+fileExistsRulesFast :: (NormalizedFilePath -> Action Bool) -> Rules ()
+fileExistsRulesFast getFileExists = do
+  addIdeGlobal . FileExistsMapVar =<< liftIO (newVar [])
+  defineEarlyCutoff $ \GetFileExists file -> do
+    fileExistsMap <- getFileExistsMapUntracked
+    let mbFilesWatched = Map.lookup file fileExistsMap
+    case mbFilesWatched of
+      Just fv -> pure (createKey fv, ([], Just fv))
+      Nothing -> do
+        exist                   <- getFileExists file
+        ShakeExtras { eventer } <- getShakeExtras
+
+        -- add a listener for VFS Create/Delete file events,
+        -- taking the FileExistsMap lock to prevent race conditions
+        -- that would lead to multiple listeners for the same path
+        modifyFileExistsAction $ \x -> case Map.lookup file x of
+          Just{}  -> return x
+          Nothing -> do
+            addListener eventer file
+            return x
+        pure (createKey exist, ([], Just exist))
+ where
+  createKey = Just . BS.toStrict . encode
+  addListener eventer fp = do
+    let
+      req = RequestMessage "2.0" reqId ClientRegisterCapability regParams
+      reqId        = IdString fpAsId
```

Comment:
Using the filepath as an id is perhaps not the best thing to do, but I could not figure out how to get access to the `IO LspId` command. Maybe it could be added to `ShakeExtras` ?